### ISSUE TITLE
Problems in setting up latex2unicode

### ISSUE THREAD

**Issue by lenhard** [tomtung/latex2unicode#1]
Thank you for providing this quite nice looking library!

For the JabRef reference manager (https://github.com/JabRef/jabref), we are currently looking for a replacement for our custom internal conversion from LaTeX to unicode. This library seems very suited and so I was giving it a try. Unfortunately, I partly get incorrect results and therefore I wonder if I did some mistake in the setup.

I added latex2unicode via a compile dependency in gradle (which by itself works fine). From the gradle file:
```    
compile 'com.github.tomtung:latex2unicode_2.11:0.1-SNAPSHOT'
```
I can use `DefaultLatexToUnicodeConverter` in our code, but it fails to perform a correct conversion in some very simple cases. For instance, the following JUnit test:
```
    @Test
    public void testUmlaut() {
        assertEquals("ä", DefaultLatexToUnicodeConverter.convert("{\\\"{a}}"));
    }
``` 
results in 
```
org.junit.ComparisonFailure: 
Expected :ä
Actual   :"a
```
Now, I would be very surprised if the conversion for something as simple as an Umlaut did not work. Also, it works fine in your web interface. Might this be a problem with the String escaping in Java? In the above example, I have to used three `\`. One to escape the following backslash and one to escape the `"`. If this is the problem, is there a way to circumvent this on the side of latex2unicode? Or will I have to manually unescape Java Strings before I pass them into latex2unicode? From the README I was expecting that escaping is no problem. Otherwise, could this be related to the encoding (UTF-8 in my case)?

Any hints are welcome!

Regards
Jörg

Refs https://github.com/JabRef/jabref/issues/2465

**Comment by tomtung** [comment]
lol can't believe someone is trying to use this thing from so long ago... I haven't even been working on JVM for years :)

Following my own instructions, I installed sbt, specified the dependency in a `build.sbt` file, and started a Scala REPL from sbt. Things seems to be working fine (with the string that you provided as test case):

```
scala> import com.github.tomtung.latex2unicode._
import com.github.tomtung.latex2unicode._

scala> val latex = "{\\\"{a}}"
latex: String = {\"{a}}

scala> val unicode = LatexToUnicodeConverter.convert(latex)
unicode: String = ä
```

This works in the [online demo](http://latex2unicode.herokuapp.com/), too.

I'm not familiar with Gradle, not sure if the dependency setting is correct.

Let me know if there's anything I can help with!

**Comment by lenhard** [comment]
Hehe, that is what you get from hosting the project on github ;-) Thank you very much for replying, even if you are not into this any more!

The error is really weird and seems to come from the fact that the Java String representation is not unicode hexes and vice versa. Let me illustrate the problem with a screenshot from a debugging session in my IDE (Intellij):

![unicode](https://cloud.githubusercontent.com/assets/1515701/22548233/ce5f8f58-e945-11e6-9f4e-0a8da0dca2d7.png)

So the result of the conversion as Java sees it (upper right corner) is not precisely the letter `ä`, but on the other hand parts of the IDE have no problem converting the result into the correct letter. I am not yet sure what is missing so that Java follows that interpretation, but it seems that one step of conversion is missing.

To add to the confusion, this working just fine:
```
assertEquals("\uD835\uDC61\uD835\uDC52\uD835\uDC65\uD835\uDC61", DefaultLatexToUnicodeConverter.convert("\\textit{text}"));
```
but this fails:
```
assertEquals("\u00E4", DefaultLatexToUnicodeConverter.convert("{\\\"{a}}"));
```
So it does not seem to be the case that latex2unicode just replies with unicode hexes.

To sum up, it seems that GUI frameworks can figure out how to display converted text correctly, but I need to find a way to get this into the java string representation exactly, since I am doing further processing on it. I have tried a few ways, but am out of ideas at the moment.

**Comment by tomtung** [comment]
One thing worth noticing is that I represented umlaut using the combining character U+0308 ◌̈ , not as individual characters like U+00E4. See https://github.com/tomtung/latex2unicode/blob/master/src/main/scala/com/github/tomtung/latex2unicode/helper/Unary.scala#L48

This could be why the assertion failed. Maybe junit is just not printing U+0308 properly.

**Comment by lenhard** [comment]
Yes, this is probably the reason for the failure of the test. On a side note, there are a few other things that fail (apart from letters with umlauts). They all relate to different versions of the letters of the english alphabet, e.g. acutes, carons, or tilde. I'll keep looking for a simple way around this.

I guess I could implement my own `LatexParser` and overwrite some methods, but then I would be re-implementing a significant part of the conversion.

**Comment by tomtung** [comment]
Yeah for simplicity I generally favored using combining characters over precomposed characters. Didn't expect this to be a problem for modern systems... Can't these characters render correctly in your GUI?

**Comment by lenhard** [comment]
Yes, it is strange that even in 2017 we seem to be fighting with the same issues that people had in the eighties. One could have thought that encodings just work these days.

The problem is not the GUI, which seems to present the characters just fine. We need the LaTeX free version for several other things, most importantly search functionality. For instance, I want to be able to search for data with my first name by typing `Jörg` into the search box, even though it is only written as `J\"{o}rg` in the data. Therefore we need to compute a LaTeX-free version of Strings internally. With the conversion problems discussed here, this will fail to match if Java thinks that precomposed characters are different from combined characters.

**Comment by tomtung** [comment]
In that case maybe you want to consider [normalizing](https://docs.oracle.com/javase/tutorial/i18n/text/normalizerapi.html) the converter output as well as your queries?

**Comment by lenhard** [comment]
Perfect, just perfect! I was not aware of this class and using it fixes most test failures. There are a few edge cases left that might have to do with our custom converter and that we could just as well change:    
- `\spreadlips{a}` seems to be an unknown command, but then again it is package-specific and imho it is sufficient to support the conversion of standard commands only. 
- So far, we just stripped unknown commands of special characters, so `\aaaaa` would turn into `aaaaa`. This is not really correct, and we might as well remove the command completely during the conversion, as latex2unicode does.

The only debatable thing left is the conversion of `32~{mA}` which converts to `32mA` and not `32 mA`. My tex compiler produces the latter, so I would say this is an error in latex2unicode. Is there a chance of repairing this and updating the package? :-)

I will round of the implementation in JabRef and then do some performance testing, but it looks like this is going to work now.

**Comment by lenhard** [comment]
I have bad news unfortunately. I did some performance benchmarking and it seems that our current ugly dirty-hack conversion solution is two orders of magnitude stronger than latex2unicode: https://github.com/JabRef/jabref/issues/2465#issuecomment-277231974. The drop is also noticeable from a user's perspective and it would render JabRef unusable for some of our power users who have huge data sets. 

I really would like to use latex2unicode since it is better in terms of correctness and it is also easy to use. However, I cannot think of a way that will magically speed it up. Maybe all the Scala/Java conversions are a problem, I do not know. In the current status, we cannot use the library.

But in any case, thank you very much for your support, this is greatly appreciated!

**Comment by tomtung** [comment]
I think it's probably due to parboiled being slow (e.g. [parboiled2](https://github.com/sirthias/parboiled2) and [fastparse](http://www.lihaoyi.com/fastparse/) seem to be faster by an order of magnitude or more). Maybe the grammar I defined could be better optimized too.

If you are not in a hurry, maybe I can figure something out to speed things up in the coming weekends.

**Comment by lenhard** [comment]
@tomtung That sounds excellent! We are not in a hurry. The replacement of our unicode conversion logic is a long term goal that I finally decided to take on, but there is no immediate need.

In this case, I will just wait for your feedback and will be more than happy to try a new and optimized version.

**Comment by tomtung** [comment]
In the meantime, it'd be really helpful if you can give me a sample set of strings that you want to parse, and some general requirements on how fast parsing needs to be :)

**Comment by lenhard** [comment]
I'll see what I can do. Would it be ok if I send you a BibTeX file? We generally just convert the content of BibTeX fields, but it wouldn't hurt if the whole file is just scanned line by line and everything is converted. 

Regarding performance requirements: The conversion is the single most important factor that determines the startup of the program, since there the content of all author, title, and some more fields are converted. So if the conversion takes 30 seconds that is how long the user has to wait until something happens after he clicked the icon. I'll try to be more conrete when I have the data.

**Comment by tomtung** [comment]
I don't really have code for extracting bibtex fields readily available... It'd be really nice it's something simpler like a text file with one string-to-parse per line.

**Comment by lenhard** [comment]
Ok I have a file ready (not a bibfile, but just the relevant strings). You can parse this file line by line and convert the result for testing. I would just need an email address so I can send you a link to it. The file contains data from one of our users and therefore I cannot make it public (and you should not either), this is just for internal testing. So please tell me your email adress or alternatively send me a mail to joerg.lenhard@kau.se then I will reply.

On my current machine converting all lines in the file in memory (just converting, reading the lines from disk has been done before) takes a little less than 500 milliseconds. Reaching this level would be desirable.

**Comment by tomtung** [comment]
@lenhard I switched to fastparse, and seems like it might be fast enough now

```scala
scala> :paste
// Entering paste mode (ctrl-D to finish)

import com.github.tomtung.latex2unicode._
import fastparse.all._

val text = "This program translates LaTeX markup to \nhuman-readable Unicode when possible.\n\nBasic math notations:\\\\\n\\because \\t{AB} + \\t{BC} \\neq \\t{AC}\n\\therefore \\iint\\sqrt[4]{\\xi^{\\theta + 1}\n- \\frac 38} \\le \\Sigma \\zeta_i \n\\therefore \\exists{x}\\forall{y} x \\in \\^A\n\nEasily type in hundreds of other symbols\nand special characters: \\spadesuit, \\aleph,\n\\OE, \\downdownarrows, etc.\n\nFont styles support: \\textbb{Black\nBoard Bold}, \\textfrak{Fraktur}, {\\bf\nBold Face},{\\cal Calligraphic},{\\it Italic},\n{\\tt Monospace}."
println(LaTeX2Unicode.convert(text))
println()

def time[R](block: => R): R = {
  val t0 = System.nanoTime()
  val result = block
  // call-by-name
  val t1 = System.nanoTime()
  println("Elapsed time: " + (t1 - t0) / 1000000000.0 + "s")
  result
}

val lines = scala.io.Source.fromFile("/Users/yubing/Downloads/testdata.txt").getLines.toArray
println(s"${lines.length} lines in test data")
time {
  for ((line, i) <- lines.zipWithIndex) {
    if(LaTeX2Unicode.parse(line).isInstanceOf[Parsed.Failure]) {
      println(s"- failed to parse line $i")
    }
  }
}

// Exiting paste mode, now interpreting.
```
```
This program translates LaTeX markup to human-readable Unicode when possible.

Basic math notations:
 ∵ A͡B + B͡C ≠ A͡C ∴ ∬∜ξ̅ᶿ̅ ̅⁺̅ ̅¹̅ ̅-̅ ̅⅜̅ ≤ Σ ζᵢ ∴ ∃x∀y x ∈ Â

Easily type in hundreds of other symbols and special characters: ♠, ℵ, Œ, ⇊, etc.

Font styles support: 𝔹𝕝𝕒𝕔𝕜 𝔹𝕠𝕒𝕣𝕕 𝔹𝕠𝕝𝕕, 𝔉𝔯𝔞𝔨𝔱𝔲𝔯, 𝐁𝐨𝐥𝐝 𝐅𝐚𝐜𝐞,𝓒𝓪𝓵𝓵𝓲𝓰𝓻𝓪𝓹𝓱𝓲𝓬,𝐼𝑡𝑎𝑙𝑖𝑐, 𝙼𝚘𝚗𝚘𝚜𝚙𝚊𝚌𝚎.

10126 lines in test data
- failed to parse line 277
- failed to parse line 3645
Elapsed time: 0.599862262s
```

**Comment by lenhard** [comment]
Great to hear! I will run the benchmarks in JabRef and then we will see.

However, are you sure the new snapshot is uploaded to the repos? So far, I can only see the old one here https://oss.sonatype.org/content/repositories/snapshots/com/github/tomtung/latex2unicode/ and my environment fails to resolve `latex2unicode_2.12:0.2-SNAPSHOT`

**Comment by tomtung** [comment]
hmm https://oss.sonatype.org/content/repositories/snapshots/com/github/tomtung/latex2unicode_2.12/ is clearly there... what error message are you getting? Maybe also try `latex2unicode_2.11:0.2-SNAPSHOT` if you are not on Java 8.

**Comment by lenhard** [comment]
My bad, the download seems to have worked fine. I was just desperately looking for the class `Latex2Unicode`, which of course did not exist in any of the loaded libraries and should have been `LaTeX2Unicode`.

So I ran some another benchmark. With the new version we are up at 72417.919 ops/s, which is a gigantic boost from 1936.683 ops/s I got for v2.11, although not at the same level as our current converter with 100036.407 ops/s. However, I do no longer have a noticeable delay in opening a database and considering the much more complete conversion that we have with your library, I am for going live with this. I'll discuss with the other developers and need to get their consent of course, but I am very positive towards this.

Thank you for your great work :)

**Comment by tomtung** [comment]
Glad to help :) I kind of wrote it in a hurry, so let me know if found any bug
